[HUDI-7779] Guard archival on savepoint removal until cleaner is able to clean it up

lokeshj1703 opened the pull request:
### Change Logs

The PR adds changes such that:
If cleaner last clean commit tracks savepointed instants and there is a replace commit following the savepoint and before earliest instant to retain for cleaner, then the archival will block on the replace commit instead of the earliest instant to retain. 
This is required to handle the issue of duplicates. If replace commit is replacing file groups retained by the savepoint, the cleaner must be able to clean these file groups before the replace commit is archived.

### Impact

NA

### Risk level (write none, low medium or high below)

low

### Documentation Update

NA

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

danny0405 commented:
@nsivabalan Is this the fix we talked before for savepoint deletion? Do we introduce cyclic dependency for clean and archiving.

jonvex commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/timeline/HoodieTimelineArchiver.java:
```diff
@@ -362,6 +367,25 @@ private boolean deleteArchivedInstants(List<ActiveAction> activeActions, HoodieE
+        String cleanerEarliestInstantToNotArchive = CleanerUtils.getCleanerPlan(metaClient, cleanInstant.isRequested()
+                ? cleanInstant
+                : HoodieTimeline.getCleanRequestedInstant(cleanInstant.getTimestamp()))
+            .getExtraMetadata().getOrDefault(CleanPlanner.EARLIEST_COMMIT_TO_NOT_ARCHIVE, null);
```
getExtraMetadata() can return null and will NPE here. Check if the map exists before doing getOrDefault

lokeshj1703 replied:
Addressed

nsivabalan reviewed (approved):
but lets wait to get an approval from Danny. 
@danny0405 : we are not introducing a cycling dependency.  Archival will try to guard up until cleaner has run. and not go beyond that. 
for eg, if we disable cleaner for 2 days, w/o this patch, archival might potentially make progress. but after this patch, there won't be any progress from archival standpoint if cleaner has not make any progress. 

try going over the PR description. I can help explain f2f if that helps.

danny0405 commented:
> Whose value will either refer to first savepoint if first savepoint < earliest commit to retain.

Does this mean we can not archive beyond any savepoint now?

> Last completed clean instant is required in the timeline to fetch earliest commit to not archive.

I don't think this is reasonable in high level, there are some scenarios that people will never enable cleaning service, or the cleaning just got stuck and the halt of archival would impact the performance of ingestion.

nbalajee commented:
Let's call out that for replacecommit (clustering operation with replace fileIDs), archiving a replacecommit before the replaced files are cleaned, would result in a data quality issue for the query engines.   Queries would include the replaced fileIDs (considering them as valid, as they are archived), which may manifest as duplicates in query results.

lokeshj1703 commented:
I have explained the scenario in the comments. PTAL @nbalajee @danny0405 
I have also fixed the earliestCommitToNotArchive logic to take into account last clean instant and removed separate handling in the cleaner timeline archival. The case where earliestCommitToNotArchive > last clean instant will not be possible since this timestamp is generated by cleaner but still added a check to consider the minimum of last clean instant and earliestCommitToNotArchive.
cc @nsivabalan

nsivabalan commented:
@danny0405 @lokeshj1703 : are we aligned on any optimization we need on top of the patch? for eg, we were discussing about replace commit timeline right? did you guys sync up on that. if not, can we sync up and make forward progress.

danny0405 commented:
> did you guys sync up on that. if not, can we sync up and make forward progress.

I thought we have synced up and reached concensus in the last stand-up meeting, no?

lokeshj1703 commented:
I have added the optimisation. @danny0405 @nsivabalan PTAL.

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/timeline/HoodieTimelineArchiver.java:
```diff
@@ -264,20 +268,46 @@ private List<HoodieInstant> getCommitInstantsToArchive() throws IOException {
+    // Archival is blocked at c2 since there is a savepoint at c2.
+    // Lets say the savepoint at c2 is now deleted, Archival would archive c3.rc since it is unblocked now.
+    // Since c3 is archived and f2 has not been cleaned, the table view would be considering f2 as a valid
+    // file id. This causes duplicates.
```
Can we put the logic inside `getEarliestInstantToRetainForClustering` to make the logic around replace_commit more concentrated and there is no need to go through the replace_commit timeline multiple times?

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanActionExecutor.java:
```diff
@@ -149,17 +150,24 @@ HoodieCleanerPlan requestClean(HoodieEngineContext context) {
+        extraMetadata.put(SAVEPOINTED_TIMESTAMPS, savepointedTimestamps.stream().collect(Collectors.joining(",")));
+      }
+      if (earliestCommitToNotArchive.isPresent()) {
+        extraMetadata.put(EARLIEST_COMMIT_TO_NOT_ARCHIVE, earliestCommitToNotArchive.get());
```
Can we put the name more straight-forward like `SKIPPED_SAVEPOINTED_COMMITS`

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java:
```diff
@@ -553,13 +552,62 @@ public Pair<Boolean, List<CleanFileInfo>> getDeletePaths(String partitionPath, O
+          earliestCommitToNotArchive = earliestToRetain.get().getTimestamp();
+        }
+      }
+      earliestCommitToNotArchiveOpt = Option.ofNullable(earliestCommitToNotArchive);
```
Can we use a separate variable like `skippedReplacedSavepoints` apart from the `earliestToRetainTimestamp` ? And let the archiver to decide which one to use.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieTimeline.java:
```diff
@@ -437,6 +437,15 @@ static boolean compareTimestamps(String commit1, BiPredicate<String, String> pre
+    return GREATER_THAN.test(commit1, commit2) ? commit2 : commit1;
+  }
+
   /**
```
Should we reuse `minInstant` instead?

lokeshj1703 commented:
@danny0405 I have updated the PR to address your comments. Can you please check?

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanActionExecutor.java:
```diff
@@ -149,17 +150,24 @@ HoodieCleanerPlan requestClean(HoodieEngineContext context) {
-      return Collections.singletonMap(SAVEPOINTED_TIMESTAMPS, savepointedTimestamps.stream().collect(Collectors.joining(",")));
+      Map<String, String> extraMetadata = new HashMap<>();
+      if (!savepointedTimestamps.isEmpty()) {
+        extraMetadata.put(SAVEPOINTED_TIMESTAMPS, savepointedTimestamps.stream().collect(Collectors.joining(",")));
```
Because we already remember the savepointed timestamps(let's name it as `s1`) and `earliestCommitToRetain` in the clean metadata, in archiver,
let's just check all the instants(name it as `s2`) within the range [min(s1), earliestCommitToRetain], if `s2` contains any replace_commits, then guard the archival with min(s1), otherwise we are good to go.

So there is no need to put any extra metadata in the clean metadata actually.

lokeshj1703 replied:
Addressed

lokeshj1703 commented on hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java:
```diff
@@ -376,9 +381,75 @@ public static Option<HoodieInstant> getEarliestInstantToRetainForClustering(
+      // When archive beyond savepoint is enabled, we do not block the archival based on cleaner earliestSavepoint
+      return Option.empty();
+    }
+    // explicitly check the savepoint timeline and guard against the first one
+    Option<String> firstSavepointOpt = activeTimeline.getSavePointTimeline().filterCompletedInstants().firstInstant().map(HoodieInstant::getTimestamp);
+    String earliestSavepointTs = firstSavepointOpt.orElse(null);
+    Option<String> cleanerEarliestSavepoint = Option.empty();
+    Option<HoodieActionInstant> cleanerEarliestToRetain = Option.empty();
+
+    if (cleanerPolicy != HoodieCleaningPolicy.KEEP_LATEST_FILE_VERSIONS && cleanInstantOpt.isPresent()) {
```
@nsivabalan Is specific handling for `KEEP_LATEST_FILE_VERSIONS` still required or can we keep logic same for all cleaner policies? It seems like SAVEPOINTED_TIMESTAMPS would always be created.

danny0405 commented:
@lokeshj1703 Thanks for the contribution, I have reviewed and created a patch: 
[7779.patch.zip](https://github.com/user-attachments/files/16276593/7779.patch.zip)

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java:
```diff
@@ -333,28 +335,31 @@ public static List<HoodieInstant> getPendingClusteringInstantTimes(HoodieTableMe
+      HoodieActiveTimeline activeTimeline, HoodieTableMetaClient metaClient, HoodieCleaningPolicy cleanerPolicy, boolean shouldArchiveBeyondSavepoint) throws IOException {
     Option<HoodieInstant> oldestInstantToRetain = Option.empty();
     HoodieTimeline replaceOrClusterTimeline = activeTimeline.getTimelineOfActions(CollectionUtils.createSet(HoodieTimeline.REPLACE_COMMIT_ACTION, HoodieTimeline.CLUSTERING_ACTION));
+    Option<HoodieInstant> cleanInstantOpt =
```
Why moves the `cleanInstantOpt` out from the check `!replaceOrClusterTimeline.empty()` ?

lokeshj1703 replied:
Addressed

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java:
```diff
@@ -365,20 +370,56 @@ public static Option<HoodieInstant> getEarliestInstantToRetainForClustering(
         oldestInstantToRetain = replaceOrClusterTimeline.firstInstant();
       }
     }
+    if (!shouldArchiveBeyondSavepoint) {
```
It looks like this chagne is not necessary if we do not make changes to lines L276 ~ L280 of `HoodieTimelineArchiver: `.

lokeshj1703 replied:
Addressed

nsivabalan commented:
hey @lokeshj1703 : can you update the PR desc

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java:
```diff
@@ -365,20 +370,56 @@ public static Option<HoodieInstant> getEarliestInstantToRetainForClustering(
     return oldestInstantToRetain;
   }
 
+  public static Option<String> getEarliestReplacedSavepointInClean(HoodieActiveTimeline activeTimeline, HoodieCleaningPolicy cleanerPolicy,
```
we can name this method as 
getEarliestCommitToNotArchive. and keep this method in HoodieTimelineArchiver. 

either we can compute and pass the value to ClusteringUtils.getEarliestInstantToRetainForClustering  or make a function as an arg.

danny0405 replied:
`getEarliestCommitToNotArchive` is not a good name because all the logic in archiver is to compute the "earliest commit to not archive".
